eth/protocols/snap/sync.go

```diff
@@ -2899,3 +2900,78 @@ func (s *capacitySort) Swap(i, j int) {
 	s.ids[i], s.ids[j] = s.ids[j], s.ids[i]
 	s.caps[i], s.caps[j] = s.caps[j], s.caps[i]
 }
+
+// healRequestSort implements the Sort interface, allowing sorting trienode
+// heal requests, which is a prerequisite for merging storage-requests.
+type healRequestSort struct {
+	hashes []common.Hash
+	paths  []trie.SyncPath
+}
+
+func (t *healRequestSort) Len() int {
+	return len(t.hashes)
+}
+
+func (t *healRequestSort) Less(i, j int) bool {
+	a := t.paths[i]
+	b := t.paths[j]
+	switch bytes.Compare(a[0], b[0]) {
+	case -1:
+		return true
+	case 1:
+		return false
+	}
+	// identical first part
+	if len(a) < len(b) {
+		return true
+	}
+	if len(b) < len(a) {
+		return false
+	}
+	if len(a) == 2 {
+		return bytes.Compare(a[1], b[1]) < 0
+	}
+	return false
+}
+
+func (t *healRequestSort) Swap(i, j int) {
+	t.hashes[i], t.hashes[j] = t.hashes[j], t.hashes[i]
+	t.paths[i], t.paths[j] = t.paths[j], t.paths[i]
+}
+
+// Merge merges the pathsets, so that several storage requests concerning the
+// same account are merged into one, to reduce bandwidth.
+// OBS: This operation is moot if t has not first been sorted.
+func (t *healRequestSort) Merge() []TrieNodePathSet {
+	var (
+		pathsets []TrieNodePathSet
+		last     TrieNodePathSet
+	)
+
+	for _, path := range t.paths {
+		pathset := TrieNodePathSet([][]byte(path))
+		if len(path) == 1 {
+			pathsets = append(pathsets, pathset)
+			last = nil
+			continue
+		}
+		if last == nil || !bytes.Equal(pathset[0], last[0]) {
+			pathsets = append(pathsets, pathset)
+			last = pathset
+			continue
+		}
+		// we can merge this into the last one
+		last = append(last, pathset[1])
```

Comment:
Question, @fjl or @karalabe . The pathset here is the same underlying slice as the `path` --  `TrieNodePathSet([][]byte(path))([][]byte(path)`. Doing `append` like this, is there any risk of somehow this operation modifying the `path`? I thought there wouldn't be, but I'd like a second opinion.

Reply:
I've looked at this specific thing quite a bit when reviewing.

There are 2 "quirky" uses:

- The append(pathset[1]) is surely safe, because even if the slice / underlying buffer is reused, `pathset[1]` already extracts an item from the buffer, so even if the **slot** is overwritten, the value originally stored in the slot should not be changed.
- The more interesting clause is `last = pathset`, since that will keep hold of the buffer. Running it in a playground seems to say it's fine, even without the `pathset = path` closure: https://go.dev/play/p/l3yE522jpxV. My 2c is that a slice is a *value*, so it is inherently a closure, just as if we were to iterate over ints or strings.

That said, our code contains a *whole* lot of closure creations for addresses and hashes... which should not be needed if my logic above holds true. So... someone's wrong, either my past of present self.

Reply:
Ah ok, in our code (statedb), we didn't save a `hash`, rather saved a `hash[:]`, which kept a reference to the underlying array. So seemingly, that's not allowed. It's different than from the current code.

Still, doens't answer whether `last = pathset` is safe or not in that case.

Reply:
According to https://go.dev/play/p/wohnnctMHgu

For value types, saving the outer variable as well as indexing into it is safe, as both creates a new value / copy.

On the other hand, slicing into is is **not** safe, as that doesn't copy, rather pulls up a reference to the internals that get changed.

Reply:
Weirdly, if I replace the `hash` definition above from `[2]byte` to `[]byte`, then all of a sudden slicing becomes safe too: https://go.dev/play/p/4LLaymk4Si7

Seems Go can't optimize the variable loop any more to avoid a copy internally? Messy messy messy.

Reply:
... So the TLDR is that you think it should be safe?

Reply:
TL;DR is let's pick @fjl's brain on what the spec says :) I think it's safe, but I'd like to know why.